### ISSUE TITLE
Model auto-update on `predict`?

### ISSUE THREAD

**Issue by justhalf** [deeplearning4j/deeplearning4j#686]
I'm following the Iris example as explained in [this documentation](http://deeplearning4j.org/iris-flower-dataset-tutorial.html).

However, it seems that each call to `predict` (which calls `output`) changes the weight of the neural network, and also the prediction. Sample code (copy the relevant code from [this file](https://raw.githubusercontent.com/deeplearning4j/dl4j-0.4-examples/master/src/main/java/org/deeplearning4j/examples/deepbelief/DBNIrisExample.java), then add the following):

```java
for(int i=0; i<10; i++){
	log.info("Output: {}", model.predict(test.getFeatureMatrix()));
}
```

produces the following:

```
Output: [2, 2, 2, 2, 1, 2, 1, 2, 0, 2, 1, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 2, 1, 2, 2, 2]
Output: [1, 1, 1, 1, 2, 2, 1, 1, 2, 2, 1, 1, 1, 1, 2, 2, 1, 2, 2, 2, 2, 2, 1, 1, 2, 2, 2, 2, 2, 1]
Output: [1, 1, 2, 2, 0, 2, 0, 1, 0, 2, 0, 2, 1, 1, 2, 0, 2, 2, 0, 2, 2, 2, 1, 1, 0, 2, 0, 2, 0, 2]
Output: [2, 2, 2, 2, 0, 2, 1, 2, 0, 0, 0, 2, 2, 2, 0, 1, 2, 2, 0, 2, 2, 0, 2, 1, 0, 2, 0, 2, 0, 2]
Output: [2, 2, 2, 2, 1, 2, 0, 2, 1, 2, 1, 2, 2, 1, 2, 0, 2, 2, 2, 2, 2, 2, 2, 1, 0, 2, 0, 2, 0, 2]
Output: [2, 2, 2, 2, 0, 2, 1, 1, 0, 0, 0, 2, 2, 1, 0, 2, 2, 2, 0, 2, 2, 0, 2, 1, 0, 2, 0, 2, 0, 2]
Output: [2, 2, 2, 2, 0, 2, 0, 2, 0, 1, 0, 2, 2, 2, 2, 0, 2, 2, 0, 2, 2, 2, 2, 2, 0, 2, 0, 2, 0, 2]
Output: [2, 2, 2, 2, 0, 2, 2, 2, 0, 0, 0, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 0, 2, 2, 0, 2, 2, 2, 2, 2]
Output: [2, 2, 2, 2, 0, 2, 0, 2, 0, 2, 0, 2, 2, 2, 2, 0, 2, 2, 0, 0, 2, 2, 2, 2, 0, 2, 0, 2, 0, 2]
Output: [2, 2, 2, 2, 0, 2, 0, 2, 0, 2, 0, 2, 2, 2, 2, 0, 2, 2, 0, 0, 2, 2, 2, 2, 0, 2, 0, 2, 0, 2]
```

**Is this an expected behavior?** I personally don't expect a model to change whenever I call a `predict` function. I can't seem to fix this issue (to have predict function that doesn't update the weights) after twiddling around with various arguments.

Also, there are inconsistencies of argument in the sequence of method calls:

```java
MultiLayerNetwork.output(INDArray input) -> output(input, false)
MultiLayerNetwork.output(INDArray input, boolean test) -> feedforward(input, test)
MultiLayerNetwork.feedForward(INDArray input, boolean test) -> feedforward(test)
MultiLayerNetwork.feedForward(boolean test) -> activationFromPrevLayer(i, currInput, test)
MultiLayerNetwork.activationFromPrevLayer(int curr, INDArray input, boolean training)
```

Notice that in the last call the boolean variable `test` is used as an argument for the boolean argument named `training` in `activationFromPrevLayer`. Is this related?

I'm using version **0.4-r1.2** from Maven Central for all dl4j and nd4j packages.

**Comment by agibsonccc** [comment]
So this is the predict method: https://github.com/deeplearning4j/deeplearning4j/blob/3d1e979bd3981770a664b59a33b07b9017a9901b/deeplearning4j-core/src/main/java/org/deeplearning4j/nn/multilayer/MultiLayerNetwork.java#L1291-1291

None of those are updating..

The testing scheme here is for whether it should be train/test. I'm going to put this off till next release (I mainly tell people to use output ;/)

I also know for a fact the weights aren't updating during predict. I think some things are being conflated. This is a lot to unravel and there's a lot of this we're not seeing on our side. I will have to either ask you for a simple unit test or put this off till I get time to figure out what's noise and what's signal.

The API changes should be a separate issue and perhaps a doc issue? I'd like to maybe split this up a bit.

**Comment by justhalf** [comment]
I ran further experiments, and yes, the weights don't change, but the result of `output` still changes. Perhaps my understanding of how a neural network should work is flawed, but I expect if the weights stay the same, the output should stay the same for the same input. Otherwise it is useless for anything.

As for the unit test, it is simply the sample given in the documentation.

I copy and simplify here for reference:

```java
import java.util.Collections;
import java.util.Random;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.deeplearning4j.datasets.iterator.DataSetIterator;
import org.deeplearning4j.datasets.iterator.impl.IrisDataSetIterator;
import org.deeplearning4j.nn.api.OptimizationAlgorithm;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.deeplearning4j.nn.conf.Updater;
import org.deeplearning4j.nn.conf.layers.OutputLayer;
import org.deeplearning4j.nn.conf.layers.RBM;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.nn.weights.WeightInit;
import org.deeplearning4j.optimize.api.IterationListener;
import org.deeplearning4j.optimize.listeners.ScoreIterationListener;
import org.nd4j.linalg.dataset.DataSet;
import org.nd4j.linalg.dataset.SplitTestAndTrain;
import org.nd4j.linalg.factory.Nd4j;
import org.nd4j.linalg.lossfunctions.LossFunctions;

public class Test {
	
	public static final Logger LOGGER = LogManager.getLogger(Test.class);
	
	public static void main(String[] args){
		final int numRows = 4;
		final int numColumns = 1;
		int outputNum = 3;
		int numSamples = 150;
		int batchSize = 150;
		int iterations = 1000;
		int splitTrainNum = (int) (batchSize * .8);
		int seed = 123;
		int listenerFreq = iterations/5;
		Nd4j.getRandom().setSeed(seed);
		LOGGER.info("Load data....");
		DataSetIterator iter = new IrisDataSetIterator(batchSize, numSamples);
		DataSet iris = iter.next();
		iris.normalizeZeroMeanZeroUnitVariance();

		LOGGER.info("Split data....");
		SplitTestAndTrain testAndTrain = iris.splitTestAndTrain(splitTrainNum, new Random(seed));
		DataSet train = testAndTrain.getTrain();
		DataSet test = testAndTrain.getTest();
		Nd4j.ENFORCE_NUMERICAL_STABILITY = true;
		
		MultiLayerNetwork model = null;
		LOGGER.info("Build model....");
		MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(seed) // Seed to lock in weight initialization for tuning
                .iterations(iterations) // # training iterations predict/classify & backprop
                .learningRate(1e-6f) // Optimization step size
                .optimizationAlgo(OptimizationAlgorithm.CONJUGATE_GRADIENT) // Backprop method (calculate the gradients)
                .l1(1e-1).regularization(true).l2(2e-4)
                .useDropConnect(true)
                .list(2) // # NN layers (does not count input layer)
                .layer(0, new RBM.Builder(RBM.HiddenUnit.RECTIFIED, RBM.VisibleUnit.GAUSSIAN)
                                .nIn(numRows * numColumns) // # input nodes
                                .nOut(3) // # fully connected hidden layer nodes. Add list if multiple layers.
                                .weightInit(WeightInit.XAVIER) // Weight initialization method
                                .k(1) // # contrastive divergence iterations
                                .activation("relu") // Activation function type
                                .lossFunction(LossFunctions.LossFunction.RMSE_XENT) // Loss function type
                                .updater(Updater.ADAGRAD)
                                .dropOut(0.5)
                                .build()
                                ) // NN layer type
                .layer(1, new OutputLayer.Builder(LossFunctions.LossFunction.MCXENT)
                                .nIn(3) // # input nodes
                                .nOut(outputNum) // # output nodes
                                .activation("softmax")
                                .build()
                                ) // NN layer type
                .build();
        model = new MultiLayerNetwork(conf);
        model.init();
		model.setListeners(Collections.singletonList((IterationListener) new ScoreIterationListener(listenerFreq)));

		LOGGER.info("Train model....");
		model.fit(train);

		for(int i=0; i<10; i++){
			LOGGER.info("Output: {}", model.output(test.getFeatureMatrix()).getRow(0));
		}
	}
}
```

As per your suggestion, I used `output` function, but it still gives wildly varying results over the same input:

```
10:13:32.972 [main] INFO  Test - Output: [ 0.30, 0.27, 0.43]
10:13:32.973 [main] INFO  Test - Output: [ 0.35, 0.41, 0.24]
10:13:32.973 [main] INFO  Test - Output: [ 0.32, 0.34, 0.34]
10:13:32.974 [main] INFO  Test - Output: [ 0.27, 0.24, 0.49]
10:13:32.974 [main] INFO  Test - Output: [ 0.32, 0.30, 0.38]
10:13:32.975 [main] INFO  Test - Output: [ 0.28, 0.27, 0.45]
10:13:32.976 [main] INFO  Test - Output: [ 0.31, 0.29, 0.40]
10:13:32.977 [main] INFO  Test - Output: [ 0.27, 0.24, 0.49]
10:13:32.978 [main] INFO  Test - Output: [ 0.30, 0.27, 0.43]
10:13:32.981 [main] INFO  Test - Output: [ 0.32, 0.30, 0.38]
```

Perhaps I am misunderstanding something about this, but I'm not sure why the output of `output` is variable.

When I pointed out the `training` and `test` parameter, I am trying to say that: Why is a boolean parameter meant to represent a `test` used in an argument that is meant to represent `training`? Clearly when `test=false` is passed to a function expecting a boolean value for training, it is wrong semantically, as is the case here.

**Comment by agibsonccc** [comment]
Turn dropout/dropconnect off. Not sure if you've seen those before but those inject random behavior in to training (which is why you're seeing random results) Test time should turn drop out/drop connect off. There maybe a bug I have to look in to(that's why we have that boolean knob either way though)

**Comment by agibsonccc** [comment]
https://github.com/deeplearning4j/deeplearning4j/pull/721
